### PR TITLE
neal 0.2.4 (new formula)

### PR DESCRIPTION
**Pull request by tadeuzagallo**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

I'd like to add NEAL, a new code analyser developed by Uber, to Homebrew. I've followed all the guidelines, the only issue I've found is the requirement that the repository is not "notable enough". However, that is expected, as it's a new project that hasn't been publicly announced yet, and my goal is to announce it with easy installation instructions via homebrew. I also went through the "Self-submitted stuff" in the "Acceptable Formulae" guide, and the project does meet most requirements: it's stable, it's maintained, it's used extensively at Uber and it does have a homepage with a lot of docs.

### CODE REVIEW

File: Formula/neal.rb [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  url "https://github.com/uber/NEAL/archive/v0.2.2.tar.gz"
+  sha256 "084e4de62ec027d16fa965727764d801b9ba7939fb6205705993ab36585c328c"
+
+  depends_on "camlp4" => :build
+  depends_on "ocaml" => :build
+  depends_on "ocamlbuild" => :build
+  depends_on "opam" => :build
+
+  def install
+    ENV.deparallelize
```
Comment (ilovezfs) [inline Formula/neal.rb]: This needs to be reported upstream

File: Formula/neal.rb [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    opamroot = buildpath/"opamroot"
+    opamroot.mkpath
+    ENV["OPAMROOT"] = opamroot
+
+    ENV["OPAMYES"] = "1"
+    ENV["NATIVE"] = "1"
+    ENV["LIB_PATH"] = lib
+
+    # otherwise we get `undefined symbol _clock_time` in El Capitan due to Core
+    ENV.delete("SDKROOT") if MacOS.version < :sierra
```
Comment (ilovezfs) [inline Formula/neal.rb]: This needs to be reported upstream

File: Formula/neal.rb [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+    system "opam", "init", "--no-setup"
+    system "opam", "install", "ocamlfind"
+    system "opam", "config", "exec", "--", "make", "setup-buildonly", "build"
+
+    bin.mkpath
```
Comment (ilovezfs) [inline Formula/neal.rb]: We need a `make install` target instead of this and what follows below.

**Review by ilovezfs — CHANGES_REQUESTED** [review]

File: Formula/neal.rb (hunk 3 above)
Comment (tadeuzagallo) [reply Formula/neal.rb]: I've updated it, thanks! I'll have to bump the version though, should I close this PR and open a new one called `NEAL v0.2.3`?

File: Formula/neal.rb (hunk 2 above)
Comment (tadeuzagallo) [reply Formula/neal.rb]: Sorry for the ignorance, but when you say upstream, do you mean the NEAL repo? Or NEAL's dependency?

**Comment by ilovezfs** [comment]
@tadeuzagallo did you want to refresh this with 0.2.3? You can force push.

**Comment by ilovezfs** [comment]
># install: /usr/local/lib/ocaml/nums.cma: Operation not permitted

This will need to depend on `ocaml-num`

File: Formula/neal.rb (hunk 2 above)
Comment (ilovezfs) [reply Formula/neal.rb]: we no longer set SDKROOT so this line should be able to just be removed

**Comment by tadeuzagallo** [comment]
Hum, it seems that it's actually the installation of `ocaml-num` that is failing. It tries to install `nums.cma` at `/usr/local/lib/ocaml`, but it does not have permissions. I'll try to figure out a way to fix it.

**Comment by ilovezfs** [comment]
@tadeuzagallo yes, that is why I created the `ocaml-num` formula. See https://github.com/Homebrew/homebrew-core/pull/20278 for an example.

**Comment by tadeuzagallo** [comment]
Oh, that makes sense. I had forgotten to rebase, so couldn't find the formula. It seems to be working now

File: Formula/neal.rb (hunk 1 above)
Comment (ilovezfs) [reply Formula/neal.rb]: What's the status of the parallelization bug(s)?

File: Formula/neal.rb [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+    (testpath/"Swift.rules").write <<-EOS.undent
+      rule NoForcedValues {
+        Swift::ForcedValueExpression {
+          fail("No forced unwrapping allowed")
+        }
+      }
+    EOS
+
+    shell_output("#{bin}/neal -r Swift.rules FailingTest.swift", 1)
```
Comment (ilovezfs) [inline Formula/neal.rb]: can we assert_match here?

File: Formula/neal.rb [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    shell_output("#{bin}/neal -r Swift.rules PassingTest.swift", 0)
```
Comment (ilovezfs) [inline Formula/neal.rb]: ditto

File: Formula/neal.rb [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    ENV["BIN_PATH"] = bin
+
+    system "opam", "init", "--no-setup"
+    system "opam", "install", "ocamlfind"
+    system "opam", "install", "--fake", "num"
+    system "opam", "config", "exec", "--", "make", "brew"
+  end
+
+  test do
+    (testpath/"FailingTest.swift").write <<-EOS.undent
```
Comment (ilovezfs) [inline Formula/neal.rb]: needs to use the new style heredocs <<~EOS

**Comment by tadeuzagallo** [comment]
Sorry, I don't know if you get notified by GitHub, so just in case, I have addressed your comments. Thanks for the review once again.

**Comment by ilovezfs** [comment]
@tadeuzagallo https://github.com/Homebrew/homebrew-core/pull/18842#discussion_r152982034

File: Formula/neal.rb (hunk 1 above)
Comment (tadeuzagallo) [reply Formula/neal.rb]: Fixed

**Review by ilovezfs — APPROVED** [review]

**Comment by ilovezfs** [comment]
Thanks for the :new: formula and your first contribution to Homebrew, @tadeuzagallo! :rocket:

**Comment by tadeuzagallo** [comment]
Thanks a lot of for the patience and all the reviews, I'm really happy it shipped! 😃

**Pull request closed without merging**